engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TypeCheckValueNode.java

```diff
@@ -0,0 +1,181 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConstructor;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.util.CachingSupplier;
+
+/** A node and a factory for nodes performing type checks (including necessary conversions). */
+public final class TypeCheckValueNode extends Node {
+  private @Child AbstractTypeCheckNode check;
+
+  TypeCheckValueNode(AbstractTypeCheckNode check) {
+    assert check != null;
+    this.check = check;
+  }
+
+  /**
+   * Wraps expression node with additional type check.
+   *
+   * @param original the expression node
+   * @param check node performing type check or {@code null}
+   * @return wrapper around {@code original} or directly {@code original} if there is {@code null}
+   *     check
+   */
+  public static ExpressionNode wrap(ExpressionNode original, TypeCheckValueNode check) {
+    if (check == null) {
+      return original;
+    } else {
+      return new TypeCheckExpressionNode(original, check);
+    }
+  }
+
+  /**
+   * Executes check or conversion of the value.
+   *
+   * @param frame frame requesting the conversion
+   * @param value the value to convert
+   * @return {@code null} when the check isn't satisfied and conversion isn't possible or non-{@code
+   *     null} value that can be used as a result
+   */
+  public final Object handleCheckOrConversion(
+      VirtualFrame frame, Object value, ExpressionNode expr) {
```

Comment:
What is the meaning of `expr` parameter? If it is non-null a conversion should be executed? Add docs please.

Reply:
Documented in 6539e48496 - it is the node that produces the `value` - if it is available.